File: books/grokking-algorithms/01_introduction_to_algorithms/binary-search.py
def binary_search(listItems, item):
  low = 0
  high = len(listItems) - 1

  while low <= high:
    mid = (low + high) // 2
    guess = listItems[mid]

    if (guess == item):
      return mid
    if (guess > item):
      high = mid - 1
    else:
      low = mid + 1

  return None

File: algorithms/binary-search.ts
function binarySearch<T>(listItems: T[], item: T) {
  let low = 0
  let high = listItems.length - 1

  while (low <= high) {
    const mid = Math.floor((low + high) / 2)
    const guess = listItems[mid]

    if (guess === item) {
      return mid
    }

    if (guess > item) {
      high = mid - 1
    }

    if (guess < item) {
      low = mid + 1
    }
  }

  return null
}
